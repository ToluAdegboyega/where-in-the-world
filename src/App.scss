@use "sass:math";

@tailwind base;
@tailwind components;
@tailwind utilities;

//Light colors

//Light Mode Text
$very-dark-blue-text: hsl(200, 15%, 8%);
//Light Mode Input
$dark-gray: hsl(0, 0%, 52%);
//Light Mode Background
$very-light-gray: hsl(0, 0%, 98%);

//Dark theme

//Dark Mode Text & Light Mode Elements
$white: hsl(0, 0%, 100%);
//Dark Mode Elements
$dark-blue: hsl(209, 23%, 22%);
//Dark Mode Background
$very-dark-blue-bg: hsl(207, 26%, 17%);

//General variables
$light-shadow: rgba(255, 255, 255, 0.1);
$dark-shadow: rgba(0, 0, 0, 0.1);
$dark-light-shadow: rgba(0, 0, 0, 0.05);
$none-shadow: inherit;

$light-opacity: 0.7;
$dark-opacity: 1;

$border-radius: 0.375rem;

$box-shadow: 0 0.25rem 0.75rem $dark-shadow, 0 0.2rem 0.1rem $dark-light-shadow;

$card-width: 15.875rem;

$transition: color 0.2s ease-in-out;
$transition-bg: background-color 0.2s ease-in-out;

@function transition($property) {
  @return $property 0.2s ease-in-out;
}

.theme {
  --element: #{$white};
  --background: #{$very-light-gray};
  --text: #{$very-dark-blue-text};
  --input: #{$dark-gray};
  --hover: #{$none-shadow};
  --opacity: #{$light-opacity};
  --option: #{$dark-shadow};

  &__light {
    --element: #{$white};
    --background: #{$very-light-gray};
    --text: #{$very-dark-blue-text};
    --input: #{$dark-gray};
    --hover: #{$none-shadow};
    --opacity: #{$light-opacity};
    --option: #{$dark-shadow};
  }
  &__dark {
    --element: #{$dark-blue};
    --background: #{$very-dark-blue-bg};
    --text: #{$white};
    --input: #{$white};
    --hover: #{$light-shadow};
    --opacity: #{$dark-opacity};
    --option: #{$light-shadow};
  }
}

.app {
  background-color: var(--background);
  transition: $transition-bg;
  grid-template-rows: 80px 1fr;
}

.header {
  background-color: var(--element);
  box-shadow: $box-shadow;
  transition-property: color, background-color;
  transition-duration: 0.2s;
  transition-timing-function: ease-in-out;

  h1 {
    a {
      color: var(--text);
      transition: transition(color);
    }
  }

  button {
    color: var(--text);
    transition-property: color, background-color;
    transition-duration: 0.2s;
    transition-timing-function: ease-in-out;

    svg {
      fill: var(--text);
    }
  }
}

.card {
  background-color: var(--element);
  box-shadow: $box-shadow;
  color: var(--text);
  transition-property: transform, background-color, color, opacity;
  transition-duration: 0.2s;
  transition-timing-function: ease-in-out;
  animation: fade-in 0.2s ease-in-out forwards;

  &:hover {
    transform: scale(1.04);
  }

  &__flag {
    img {
      color: transparent;
      background-color: var(--option);
      transition: transition(background-color);
    }
  }
}

.filter-search {
  .search-bar {
    background-color: var(--element);
    box-shadow: $box-shadow;
    transition: transition(background-color);

    svg {
      fill: var(--input);
      transition: transition(fill);
    }

    .close-btn {
      left: initial;
      color: var(--input);
      transition: transition(color);
    }

    input {
      z-index: 4;
      border-radius: $border-radius;
      color: var(--text);
      transition-property: color, background-color, opacity;
      transition-duration: 0.2s;
      transition-timing-function: ease-in-out;

      &::placeholder {
        color: var(--input);
        opacity: 1;
        transition: transition(color);
      }

      &:focus-visible {
        outline: none;
      }
    }
  }
}

.filter {
  background-color: var(--element);
  box-shadow: $box-shadow;
  transition: transition(background-color);

  button {
    width: 100%;
    height: 100%;
    border: none;
    border-radius: 0.375rem;
    background-color: transparent;
    cursor: pointer;
    text-align: left;
    padding-left: 24px;
    color: var(--text);
    transition-property: color, opacity, background-color;
    transition-duration: 0.2s;
    transition-timing-function: ease-in-out;
    font-size: 14px;
    line-height: 14px;

    &:hover {
      transition: none;
      opacity: var(--opacity);
      background-color: var(--hover);
    }

    &.open svg {
      transform: rotateZ(180deg);
    }

    svg {
      position: absolute;
      height: 16px;
      right: 18px;
      top: 1.25rem;
      transition: transform 0.2s ease-in-out;
    }
  }

  .options {
    transform: scaleY(0);
    opacity: 0;
    transform-origin: top;
    position: relative;
    margin-top: 4px;
    z-index: 9999999999999999999999999;
    width: 100%;
    border-radius: $border-radius;
    background-color: var(--element);
    box-shadow: $box-shadow;
    color: var(--text);
    transition-property: transform, opacity, background-color, color;
    transition-duration: 0.2s;
    transition-timing-function: ease-in-out;

    padding: 12px;

    .option {
      width: 100%;
      padding: 10px 12px;
      cursor: pointer;
      font-size: 14px;
      line-height: 14px;
      border-radius: math.div($border-radius, 2);
      transition: transition(background-color);

      &:hover {
        transition: none;
        background-color: var(--option);
      }
    }
  }

  .selected {
    transform: scaleY(1);
    opacity: 1;
  }
}

.back {
  width: 8rem;
  height: 2.5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--element);
  border-radius: $border-radius;
  box-shadow: $box-shadow;
  text-decoration: none;
  margin-bottom: 3rem;
  transition: transition(background-color);

  button {
    height: 100%;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    background: none;
    border: none;
    border-radius: $border-radius;
    font-size: 16px;
    line-height: 16px;
    position: relative;
    cursor: pointer;
    color: var(--text);
    transition-property: color, background-color, opacity;
    transition-duration: 0.2s;
    transition-timing-function: ease-in-out;

    &:hover {
      transition: none;
      opacity: var(--opacity);
      background-color: var(--hover);
    }

    svg {
      position: absolute;
      height: 1rem;
      width: 1rem;
      left: 2rem;
    }

    p {
      padding-left: 0.5rem;
    }
  }
}

.image-wrap {
  img {
    box-shadow: $box-shadow;
    color: transparent;
    background-color: var(--option);
    transition: transition(background-color);
  }
}

.text-wrap {
  color: var(--text);
  transition: transition(color);
}

.country-border {
  background-color: var(--element);
  button {
    color: var(--text);
  }
}

.pagination {
  span {
    color: var(--text);
  }
}

.no-result {
  color: var(--text);
}
