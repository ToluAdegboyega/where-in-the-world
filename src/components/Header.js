import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { ReactComponent as Moon } from "../assets/icons/moon.svg";
import { ReactComponent as MoonSolid } from "../assets/icons/moon-solid.svg";

const Header = () => {
  const light = "theme__light";
  const dark = "theme__dark";

  let html = document.documentElement;

  const [theme, setTheme] = useState(localStorage.getItem('theme') || light);

  const switchTheme = () => {
    if (theme === dark) {
      setTheme(light);
    } else {
      setTheme(dark);
    }
  };

  useEffect(() => {
    localStorage.setItem('theme', theme);
    html.className = theme;
  }, [html, theme]);

  return (
    <div className="header h-20 w-100 flex justify-between items-center py-0 md:px-20 px-6">
      <h1 className="md:text-2xl text-base font-semibold cursor-pointer">
        <Link to="/" className="px-0 py-2">
          Where In The World?
        </Link>
      </h1>

      <button
        data-testid="switch"
        onClick={switchTheme}
        className="flex justify-end items-center 
        h-12 py-0 px-4 -mr-4 text-xs font-semibold cursor-pointer border-none bg-transparent rounded-md"
      >
        {theme === light ? <MoonSolid className="h-5 m-2" /> : <Moon className="h-5 m-2" />}
        {theme === light ? "Dark" : "Light"} Mode
      </button>
    </div>
  );
};

export default Header;
