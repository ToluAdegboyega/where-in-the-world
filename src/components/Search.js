import { useEffect, useState } from "react";

import { ReactComponent as SearchIcon } from "../assets/icons/search.svg";
import { ReactComponent as Close } from "../assets/icons/close.svg";

const Search = ({ handleSearch }) => {
  const [search, setSearch] = useState("");
  const [block, setBlock] = useState(true);

  const handleChange = (e) => {
    let { value } = e.target;
    if (value === "") {
      setBlock(false);
    }
    setSearch(value);
  };

  const handleEnter = (e) => {
    let { key } = e;
    if (key === "Enter" && search) {
      handleSearch(search);
    }
  };

  const handleDelete = () => {
    setSearch("");
    setBlock(false);
  };

  useEffect(() => {
    if (search !== "") {
      const timeout = setTimeout(() => handleSearch(search), 500);
      return () => clearTimeout(timeout);
    } else if (!block) {
      setBlock(true);
      handleSearch(search);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [search, handleSearch]);

  return (
    <div className="search-bar m-4 max-w-md h-14 rounded-md flex justify-center items-center relative">
      <SearchIcon className="h-5 absolute left-8 top-4" />
      <input
        type="text"
        placeholder="Search"
        value={search}
        onChange={handleChange}
        onKeyUp={handleEnter}
        className="relative h-full w-full py-2 pr-12 pl-20 border-none tracking-wide font-normal text-sm bg-transparent"
      />
      {search && <Close className={"close-btn h-8 p-1 top-3 right-3 rounded-full cursor-pointer z-10"} onClick={handleDelete} />}
    </div>
  );
};

export default Search;