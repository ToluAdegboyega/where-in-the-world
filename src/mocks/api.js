export const getCountriesApi = async () => {
  const res = await fetch("https://restcountries.com/v2/all");
  return await res.json();
};
