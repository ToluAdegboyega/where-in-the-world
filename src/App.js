import { useEffect, useState } from "react";
import { HashRouter as Router, Route, Routes } from "react-router-dom";

import "./App.scss";

import Country from "./views/Country";
import Error from "./views/Error";
import Home from "./views/Home";
import Header from "./components/Header";
import { getCountriesApi } from "./mocks/api";

function App() {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setLoading(true);
    getCountriesApi()
    .then((json) => {
      if (json.message) {
        throw new Error(json.message);
      } else {
        setData(json);
        setLoading(false);
      }
    })
    .catch(() => {
      setError(true);
      setLoading(false);
    });
  }, []);

  return (
    <div className="app min-h-screen grid">
      <Router>
        <Header />
        {!loading ? (
          <Routes>
            <Route
              exact
              path="/"
              element={
                !error ? data && <Home data={data} /> : <Error error={error} />
              }
            />
            {data &&
              data.map((country) => {
                return (
                  <Route
                    exact
                    path={`/${country.alpha3Code}`}
                    key={country.alpha3Code}
                    element={<Country country={country} data={data} />}
                  />
                );
              })}
            <Route exact path="/:code" element={<Error error={error} />} />
          </Routes>
        ) : (
          <div data-testid="loading" className="flex justify-center items-center">
            <div className="flex h-14 w-14 items-center justify-center rounded-full bg-gradient-to-tr from-indigo-500 to-pink-500 animate-spin">
              <div className="h-9 w-9 rounded-full bg-white"></div>
            </div>
          </div>
        )}
      </Router>
    </div>
  );
}

export default App;
