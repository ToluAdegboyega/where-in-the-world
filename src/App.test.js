import { fireEvent, render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import App from "./App";
import Header from "./components/Header";
import Search from "./components/Search";
import api from "./mocks/api";

beforeEach(() => {
  jest.spyOn(window, "fetch").mockImplementation(api);
})

afterEach(() => {
  jest.restoreAllMocks()
});

test("renders landing page", async () => {
  render(<App />);
  expect(screen.getByRole("heading")).toHaveTextContent(/Where In The World?/);
  expect(<App />).toMatchSnapshot();
  await waitForElementToBeRemoved(() => screen.queryByTestId("loading"))
});

test("theme switch", () => {
  render(
    <Router>
      <Header />
    </Router>
  );

  const switchBtn = screen.getByTestId("switch");

  fireEvent.click(switchBtn);

  expect(switchBtn).toHaveTextContent(/Mode/);
  expect(<Header />).toMatchSnapshot();
});

test("should be able to search", () => {
  render(<Search />);

  expect(screen.getByPlaceholderText("Search")).toBeInTheDocument();

  const searchBtn = screen.getByTestId("search");
  fireEvent.keyUp(searchBtn);
  fireEvent.change(searchBtn);
});