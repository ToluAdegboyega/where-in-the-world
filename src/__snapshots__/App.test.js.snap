// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders landing page 1`] = `<App />`;

exports[`theme switch 1`] = `<Header />`;
